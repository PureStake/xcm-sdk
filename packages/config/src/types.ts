export type AssetId = number | bigint | string;
