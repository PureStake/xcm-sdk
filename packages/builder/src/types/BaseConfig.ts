import { CallType } from '../builder.interfaces';

export interface BaseConfigConstructorParams {
  module: string;
  func: string;
  type: CallType;
}

export class BaseConfig {
  readonly module: string;

  readonly func: string;

  readonly type: CallType;

  constructor({ module, func, type }: BaseConfigConstructorParams) {
    this.module = module;
    this.func = func;
    this.type = type;
  }
}
