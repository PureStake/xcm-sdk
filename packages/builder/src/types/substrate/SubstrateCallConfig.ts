import { ApiPromise } from '@polkadot/api';
import { CallType } from '../../builder.interfaces';

export interface SubstrateCallConfigConstructorParams {
  api: ApiPromise;
  call: () => Promise<bigint>;
}

export class SubstrateCallConfig {
  readonly api: ApiPromise;

  readonly call: () => Promise<any>;

  readonly type = CallType.Substrate;

  constructor({ api, call }: SubstrateCallConfigConstructorParams) {
    this.api = api;
    this.call = call;
  }
}
