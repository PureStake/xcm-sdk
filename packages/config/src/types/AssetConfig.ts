import {
  AssetMinConfigBuilder,
  BalanceConfigBuilder,
  ContractConfigBuilder,
  ExtrinsicConfigBuilder,
  FeeConfigBuilder,
} from '@moonbeam-network/xcm-builder';
import { AnyChain, Asset } from '@moonbeam-network/xcm-types';

export interface AssetConfigConstructorParams {
  asset: Asset;
  balance: BalanceConfigBuilder;
  contract?: ContractConfigBuilder;
  destination: AnyChain;
  destinationFee: DestinationFeeConfig;
  extrinsic?: ExtrinsicConfigBuilder;
  fee?: FeeAssetConfig;
  min?: AssetMinConfigBuilder;
}

export interface DestinationFeeConfig extends FeeAssetConfig {
  amount: number | FeeConfigBuilder;
}

export interface FeeAssetConfig {
  asset: Asset;
  balance: BalanceConfigBuilder;
  xcmDeliveryFeeAmount?: number;
}

export class AssetConfig {
  readonly asset: Asset;

  readonly balance: BalanceConfigBuilder;

  readonly contract?: ContractConfigBuilder;

  readonly destination: AnyChain;

  readonly destinationFee: DestinationFeeConfig;

  readonly extrinsic?: ExtrinsicConfigBuilder;

  readonly fee?: FeeAssetConfig;

  readonly min?: AssetMinConfigBuilder;

  constructor({
    asset,
    balance,
    contract,
    destination,
    destinationFee,
    extrinsic,
    fee,
    min,
  }: AssetConfigConstructorParams) {
    this.asset = asset;
    this.balance = balance;
    this.contract = contract;
    this.destination = destination;
    this.destinationFee = destinationFee;
    this.extrinsic = extrinsic;
    this.fee = fee;
    this.min = min;
  }
}
