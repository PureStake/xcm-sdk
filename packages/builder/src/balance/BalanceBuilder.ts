/* eslint-disable sort-keys */
/* eslint-disable @typescript-eslint/no-use-before-define */

import { u128 } from '@polkadot/types';
import { PalletBalancesAccountData } from '@polkadot/types/lookup';
import { QueryConfig } from '../QueryConfig';
import {
  BalanceConfigBuilder,
  EquilibriumSystemBalanceData,
  PalletBalancesAccountDataOld,
  TokensPalletAccountData,
} from './BalanceBuilder.interfaces';

export function BalanceBuilder() {
  return {
    assets,
    system,
    tokens,
  };
}

function assets() {
  return {
    account: (): BalanceConfigBuilder => ({
      build: ({ address, asset }) =>
        new QueryConfig({
          module: 'assets',
          func: 'account',
          args: [asset, address],
          transform: (response: any): bigint =>
            (response.balance as u128).toBigInt(),
        }),
    }),
  };
}

function system() {
  return {
    account: (): BalanceConfigBuilder => ({
      build: ({ address }) =>
        new QueryConfig({
          module: 'system',
          func: 'account',
          args: [address],
          transform: (response: any): bigint => {
            const balance = response.data as PalletBalancesAccountData &
              PalletBalancesAccountDataOld;
            const frozen = balance.miscFrozen ?? balance.frozen;

            return BigInt(balance.free.sub(frozen).toString());
          },
        }),
    }),
    accountEquilibrium: (): BalanceConfigBuilder => ({
      build: ({ address, asset }) =>
        new QueryConfig({
          module: 'system',
          func: 'account',
          args: [address],
          transform: (response): bigint => {
            if (response.data.isEmpty) {
              return 0n;
            }

            const res = response.data.toJSON() as unknown;
            let balances: EquilibriumSystemBalanceData | undefined;

            if (Array.isArray(res)) {
              balances = res;
            }

            if (Array.isArray((res as any)?.v0?.balance)) {
              balances = (res as any).v0.balance;
            }

            if (!balances) {
              throw new Error("Can't get balance from Equilibrium chain");
            }

            const balance = balances.find(([assetId]) => assetId === asset);

            if (!balance) {
              return 0n;
            }

            return BigInt(balance[1].positive);
          },
        }),
    }),
  };
}

function tokens() {
  return {
    accounts: (): BalanceConfigBuilder => ({
      build: ({ address, asset }) =>
        new QueryConfig({
          module: 'tokens',
          func: 'accounts',
          args: [address, asset],
          transform: ({ free, frozen }: TokensPalletAccountData): bigint =>
            BigInt(free.sub(frozen).toString()),
        }),
    }),
  };
}
