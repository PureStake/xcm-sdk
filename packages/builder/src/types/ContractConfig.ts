/* eslint-disable @typescript-eslint/no-explicit-any */
import { BaseConfig, BaseConfigConstructorParams } from './BaseConfig';

export interface ContractConfigConstructorParams
  extends BaseConfigConstructorParams {
  address?: string;
  args: any[];
}

export class ContractConfig extends BaseConfig {
  readonly address?: string;

  readonly args: any[];

  static is(obj: unknown): obj is ContractConfig {
    return obj instanceof ContractConfig;
  }

  constructor({ args, address, ...other }: ContractConfigConstructorParams) {
    super({ ...other });

    this.args = args;
    this.address = address;
  }
}
