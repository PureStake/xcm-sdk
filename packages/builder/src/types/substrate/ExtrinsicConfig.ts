/* eslint-disable @typescript-eslint/no-explicit-any */
import { SubmittableExtrinsicFunction } from '@polkadot/api/types';
import { BaseConfig, BaseConfigConstructorParams } from '../BaseConfig';

export interface ExtrinsicConfigConstructorParams
  extends Omit<BaseConfigConstructorParams, 'type'> {
  getArgs: (func?: SubmittableExtrinsicFunction<'promise'>) => any[];
}

export class ExtrinsicConfig extends BaseConfig {
  getArgs: (func?: SubmittableExtrinsicFunction<'promise'>) => any[];

  static is(obj: unknown): obj is ExtrinsicConfig {
    return obj instanceof ExtrinsicConfig;
  }

  constructor({ getArgs, ...other }: ExtrinsicConfigConstructorParams) {
    super({ ...other });

    this.getArgs = getArgs;
  }
}
