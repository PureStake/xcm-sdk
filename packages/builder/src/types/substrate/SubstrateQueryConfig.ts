import { SetOptional } from '@moonbeam-network/xcm-types';
import { CallType } from '../../builder.interfaces';
import { BaseConfig, BaseConfigConstructorParams } from '../BaseConfig';

export interface QueryConfigConstructorParams
  extends SetOptional<BaseConfigConstructorParams, 'type'> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  args?: any[];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  transform: (data: any) => Promise<bigint>;
}

export class SubstrateQueryConfig extends BaseConfig {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  readonly args: any[];

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  readonly transform: (data: any) => Promise<bigint>;

  constructor({
    args = [],
    transform,
    type = CallType.Substrate,
    ...other
  }: QueryConfigConstructorParams) {
    super({ ...other, type });

    this.args = args;
    this.transform = transform;
  }
}
