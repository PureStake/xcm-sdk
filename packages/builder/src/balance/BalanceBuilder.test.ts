import { TypeRegistry, U128 } from '@polkadot/types';
import { describe, expect, it } from '@jest/globals';
import { SubstrateQueryConfig } from '../types/substrate/SubstrateQueryConfig';
import { BalanceBuilder } from './BalanceBuilder';

function balanceOf(number: number | string): U128 {
  return new U128(new TypeRegistry(), number);
}

/**
 * Using snapshot to test bigint values because jest does not support bigint
 */

describe('balanceBuilder', () => {
  const account = '<ACCOUNT>';
  const asset = '<ASSET>';

  describe('assets', () => {
    describe('account', () => {
      const config = BalanceBuilder()
        .substrate()
        .assets()
        .account()
        .build({ address: account, asset }) as SubstrateQueryConfig;

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', async () => {
        await expect(
          config.transform({
            unwrapOrDefault: () => ({ balance: balanceOf(999) }),
          }),
        ).resolves.toMatchSnapshot();
      });
    });
  });

  describe('system', () => {
    describe('account', () => {
      const config = BalanceBuilder()
        .substrate()
        .system()
        .account()
        .build({ address: account, asset }) as SubstrateQueryConfig;

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly with frozen balance', async () => {
        await expect(
          config.transform({
            data: { free: balanceOf(999), frozen: balanceOf(99) },
          }),
        ).resolves.toMatchSnapshot();
      });

      it('should transform correctly with miscFrozen balance', async () => {
        await expect(
          config.transform({
            data: { free: balanceOf(999), miscFrozen: balanceOf(99) },
          }),
        ).resolves.toMatchSnapshot();
      });
    });

    describe('accountEquilibrium', () => {
      const config = BalanceBuilder()
        .substrate()
        .system()
        .accountEquilibrium()
        .build({ address: account, asset: 25969 }) as SubstrateQueryConfig;

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should return 0 if response empty', async () => {
        await expect(
          config.transform({
            data: {
              isEmpty: true,
            },
          }),
        ).resolves.toMatchSnapshot();
      });

      it('should transform v0 correctly', async () => {
        await expect(
          config.transform({
            data: {
              isEmpty: false,
              toJSON: () => ({
                v0: {
                  balance: [
                    [
                      25969,
                      {
                        positive: 2954852364913,
                      },
                    ],
                    [
                      6648164,
                      {
                        positive: 18800000000,
                      },
                    ],
                    [
                      1735159154,
                      {
                        positive: 245805795,
                      },
                    ],
                  ],
                  lock: 0,
                },
              }),
            },
          }),
        ).resolves.toMatchSnapshot();
      });

      it('should transform correctly if not v0', async () => {
        await expect(
          config.transform({
            data: {
              isEmpty: false,
              toJSON: () => [
                [
                  25969,
                  {
                    positive: 62730076931,
                  },
                ],
                [
                  1735159154,
                  {
                    positive: 44972367,
                  },
                ],
              ],
            },
          }),
        ).resolves.toMatchSnapshot();
      });
    });
  });

  describe('tokens', () => {
    describe('accounts', () => {
      const config = BalanceBuilder()
        .substrate()
        .tokens()
        .accounts()
        .build({ address: account, asset }) as SubstrateQueryConfig;

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', async () => {
        await expect(
          config.transform({ free: balanceOf(999), frozen: balanceOf(99) }),
        ).resolves.toMatchSnapshot();
      });
    });
  });
});
