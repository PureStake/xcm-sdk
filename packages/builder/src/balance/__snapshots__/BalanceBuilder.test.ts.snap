// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`balanceBuilder assets account should be correct config 1`] = `
QueryConfig {
  "args": [
    "<ASSET>",
    "<ACCOUNT>",
  ],
  "func": "account",
  "pallet": "assets",
  "transform": [Function],
}
`;

exports[`balanceBuilder ormlTokens accounts should be correct config 1`] = `
QueryConfig {
  "args": [
    "<ACCOUNT>",
    {
      "ForeignAsset": "<ASSET>",
    },
  ],
  "func": "accounts",
  "pallet": "ormlTokens",
  "transform": [Function],
}
`;

exports[`balanceBuilder system account should be correct config 1`] = `
QueryConfig {
  "args": [
    "<ACCOUNT>",
  ],
  "func": "account",
  "pallet": "system",
  "transform": [Function],
}
`;

exports[`balanceBuilder system accountEquilibrium should be correct config 1`] = `
QueryConfig {
  "args": [
    "<ACCOUNT>",
  ],
  "func": "account",
  "pallet": "system",
  "transform": [Function],
}
`;

exports[`balanceBuilder tokens accounts should be correct config 1`] = `
QueryConfig {
  "args": [
    "<ACCOUNT>",
    "<ASSET>",
  ],
  "func": "accounts",
  "pallet": "tokens",
  "transform": [Function],
}
`;
